src/useSynapse.js

```diff
@@ -1,25 +1,48 @@
-import React, { useContext, useState, useEffect } from 'react';
+import React, { useContext, useState, useEffect, useRef } from 'react';
 import Context from './Context';
 import shallowEqual from './shallow-equal';
 
-export default function useSynapse(selector) {
+export default function useSynapse(selector, dependencies = []) {
   const synapse = useContext(Context);
   const select = () => selector(synapse.stores);
   const [state, setState] = useState(select());
 
-  const updateStateIfChanged = () => {
-    const nextState = select();
-    if (!shallowEqual(nextState, state)) {
-      setState(nextState);
-    }
-  }
+  // By default, our effect only fires on mount and unmount, meaning it won't see the
+  // changes to state, so we use a mutable ref to track the current value
+  const stateRef = useRef(state);
 
   useEffect(() => {
-    updateStateIfChanged();
-    return synapse.subscribe(() => {
-      updateStateIfChanged()
-    });
-  });
+    // Helps to avoid running stale listeners after unmount
+    let isUnsubscribed = false;
+
+    const maybeUpdateState = () => {
+      if (isUnsubscribed) {
+        return;
+      }
+
+      const nextState = select();
+
+      // Checking referential equality grants perf boost if selector is memoized
+      if (
+        nextState === stateRef.current ||
+        shallowEqual(nextState, stateRef.current)
+      ) {
+        return;
+      }
+
+      stateRef.current = nextState;
+      setState(nextState);
+    };
```

Comment:
Is there any particular reason for not defining `#maybeUpdateState` outside of `useEffect` or it doesn't even matter?

Reply:
Couple reasons: 
- To access `isUnsubscribed` in closure
- It'll only be defined once instead of every time the hook is called